servlet.py

```diff
@@ -549,18 +552,19 @@ def get(self):
                     tempFile.seek(0)
 
                     mtype = self.getMimeType(tempFile.name)
-                    if mtype in allowedMimeTypes:
-                        self.request.headers['Content-Type'] = 'application/octet-stream'
-                        self.request.headers['Content-Disposition'] = 'attachment'
-                        self.write(translation.translateDoc(tempFile,
-                                                            allowedMimeTypes[mtype],
-                                                            self.pairs['%s-%s' % (l1, l2)],
-                                                            markUnknown))
-                        self.finish()
-                    else:
-                        self.send_error(400, explanation='Invalid file type %s' % mtype)
+                    with (yield self.doc_pipe_sem.acquire()):
+                        if mtype in allowedMimeTypes:
+                            self.request.headers['Content-Type'] = 'application/octet-stream'
+                            self.request.headers['Content-Disposition'] = 'attachment'
+                            self.write(translation.translateDoc(tempFile,
```

Comment:
extract the results of translateDoc into a variable, and put the "with" around only the call to translateDoc

(you may also have to use `@gen.coroutine`, not `@tornado.web.asynchronous`, as the wrapper – that's what they use at http://www.tornadoweb.org/en/stable/locks.html#semaphore )